test/e2e_env/universal/meshretry/grpc.go

```diff
@@ -0,0 +1,137 @@
+				WithServiceName("test-server"),
+				WithProtocol("grpc"),
+				WithArgs([]string{"grpc", "server", "--port", "8080"}),
+				WithTransparentProxy(true),
+			)).
+			Install(TestServerUniversal("test-client", meshName,
+				WithServiceName("test-client"),
+				WithArgs([]string{"grpc", "client", "--address", "test-server.mesh:80", "--unary", "true"}),
+				WithProtocol("grpc"),
+				WithTransparentProxy(true),
+			)).
+			Setup(env.Cluster)
+		Expect(err).ToNot(HaveOccurred())
+
+		// Delete the default retry policy
+		Eventually(func() error {
+			return env.Cluster.GetKumactlOptions().RunKumactl("delete", "retry", "--mesh", meshName, "retry-all-"+meshName)
+		}).Should(Succeed())
+	})
+
+	E2EAfterAll(func() {
+		Expect(env.Cluster.DeleteMeshApps(meshName)).To(Succeed())
+		Expect(env.Cluster.DeleteMesh(meshName)).To(Succeed())
+	})
+
+	It("should retry on GRPC connection failure", func() {
+		echoServerDataplane := fmt.Sprintf(`
+type: Dataplane
+mesh: "%s"
+name: fake-echo-server
+networking:
+  address:  241.0.0.1
+  inbound:
+  - port: 7777
+    servicePort: 7777
+    tags:
+      kuma.io/service: test-server
+      kuma.io/protocol: grpc
+`, meshName)
+		meshRetryPolicy := fmt.Sprintf(`
+type: MeshRetry
+mesh: "%s"
+name: fake-meshretry-policy
+spec:
+  targetRef:
+    kind: MeshService
+    name: test-client
+  to:
+    - targetRef:
+        kind: MeshService
+        name: test-server
+      default:
+        grpc:
+          numRetries: 5
+`, meshName)
+		admin, err := env.Cluster.GetApp("test-client").GetEnvoyAdminTunnel()
+		Expect(err).ToNot(HaveOccurred())
+
+		var lastFailureStats = stats.StatItem{Name: "", Value: float64(0)}
+		grpcFailureStats := func(g Gomega) *stats.Stats {
+			s, err := admin.GetStats("cluster.test-server.grpc.failure")
+			g.Expect(err).ToNot(HaveOccurred())
+			fmt.Printf("current failure stats %v\n", s)
+			return s
+		}
+		grpcSuccessStats := func(g Gomega) *stats.Stats {
+			s, err := admin.GetStats("cluster.test-server.grpc.success")
+			g.Expect(err).ToNot(HaveOccurred())
+			return s
+		}
+
+		By("Checking requests succeed")
+		Eventually(func(g Gomega) {
+			g.Expect(grpcSuccessStats(g)).To(stats.BeGreaterThanZero())
+		}, "30s", "1s").Should(Succeed())
+
+		Consistently(func(g Gomega) {
+			failureStats := grpcFailureStats(g)
+			if len(failureStats.Stats) != 0 {
+				defer func() { lastFailureStats = failureStats.Stats[0] }()
+				g.Expect(failureStats).To(Not(stats.BeGreaterThan(lastFailureStats)))
+			}
+		}).Should(Succeed())
+
+		By("Adding a faulty dataplane")
+		Expect(env.Cluster.Install(YamlUniversal(echoServerDataplane))).To(Succeed())
+
+		By("Check some errors happen")
+		Eventually(func(g Gomega) {
+			failureStats := grpcFailureStats(g)
+			defer func() { lastFailureStats = failureStats.Stats[0] }()
+			g.Expect(grpcFailureStats(g)).To(stats.BeGreaterThanZero())
+		}, "90s", "10s").Should(Succeed())
+		time.Sleep(10 * time.Second)
```

Comment:
We have to sleep here otherwise `Eventually` finishes and `Consistently` kicks in and in some small % of cases there is no new error occurring and the test fails on the assertion from line 119.